### PR TITLE
Refactor kReadCommissioningInfo step in DeviceCommissioner

### PR DESCRIPTION
**Pull request by ksperling-apple**
This PR refactors the kReadCommissioningInfo DeviceCommissioner step in preparation for adding the ability to read additional data. Having a fixed split of the attributes to be read into two steps in inflexible and makes it hard to optimize the reads that are performed.

Specifically this change
- Rolls kReadCommissioningInfo2 stage into kReadCommissioningInfo and tracks progress internally instead
- Groups attributes being read and parsing methods by cluster (where this wasn't already done)
- Makes the parsing code more consistent and simplifies some logic.

If desired this PR can be reviewed commit by commit; builds and tests were green after each commit.

### CODE REVIEW

File: src/controller/CHIPDeviceController.cpp [hunk 1]
```diff
@@ -2189,29 +2190,110 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
...
+void DeviceCommissioner::ContinueReadingCommissioningInfo(const CommissioningParameters & params)
+{
+    VerifyOrDie(mCommissioningStage == CommissioningStage::kReadCommissioningInfo);
+
+    // We can ony read 9 paths per Read Interaction, since that is the minimum
+    // a server has to support per spec (see "Interaction Model Limits").
+    app::AttributePathParams readPaths[InteractionModelEngine::kMinSupportedPathsPerReadRequest];
+    static_assert(InteractionModelEngine::kMinSupportedPathsPerReadRequest >= 9);
+
+    const auto proxy    = mDeviceBeingCommissioned;
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:2218]: So what happens if we were in the middle of a read and commissioning got canceled?  I think proxy here will be null, right?  

In general, the handling of things that happen during the async parts of this process needs some careful thinking about.  It was handled by the general "go to next step" machinery before, but now we have to make sure it's correct.

File: src/controller/CHIPDeviceController.cpp [hunk 2]
```diff
@@ -2916,85 +2998,23 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
...
     case CommissioningStage::kReadCommissioningInfo: {
-        ChipLogProgress(Controller, "Sending read request for commissioning information");
-        // Allocate a new ClusterStateCache when starting reading the first batch of attributes.
+        VerifyOrDie(endpoint == kRootEndpointId);
+        ChipLogProgress(Controller, "Sending read requests for commissioning information");
+
+        // Allocate a ClusterStateCache to collect the data from our read requests.
         // The cache will be released in:
         // - SendCommissioningReadRequest when failing to send a read request.
         // - ParseCommissioningInfo when the last ReadCommissioningInfo stage is completed.
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:3007]: There is no ParseCommissioningInfo anymore.

File: src/controller/CHIPDeviceController.cpp (hunk 1 above)
Comment (ksperling-apple) [reply src/controller/CHIPDeviceController.cpp:2218]: Hm, that was true for the initial `kReadCommissioningInfo` step, but I'm not sure cancellation during `kRead...2` was actually handled correctly. In `CancelCommissioningInteractions` we kill the ReadClient, and this calls `OnError` and `OnDone`. Only `OnDone` is currently implemented, and `ParseCommissioningInfo` will then probably happen to fail... Maybe `CancelCommissioningInteractions` needs to just finish the step first, and then we can just ignore the `OnDone` call.

Comment (ksperling-apple) [reply src/controller/CHIPDeviceController.cpp:2219]: Actually that's not quite correct... the ReadClient destructor won't call `OnDone`, so I don't think there is anything different to be done here about how we handle cancellation before and after my change.

File: src/controller/CHIPDeviceController.cpp [hunk 3]
```diff
@@ -2189,46 +2194,166 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
...
+    app::AttributePathParams * paths() { return mPaths; }
+
+    // Adds an attribute path if within the current window.
+    // Returns false if the available space has been exceeded.
+    template <typename... Ts>
+    bool AddAttributePath(Ts &&... args)
     {
-        err = ParseCommissioningInfo2(info);
+        VerifyOrReturnValue(mSkip == 0, true, mSkip--);
+        VerifyOrReturnValue(mCount < kCapacity, false, mCount = kCapacity + 1);
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:2236]: These should be if's . The 3 complex args are really really hard to parse on a single read.

File: src/controller/CHIPDeviceController.cpp [hunk 4]
```diff
@@ -2189,46 +2194,166 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
...
+        VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::GeneralCommissioning::Id,
+                                                Clusters::GeneralCommissioning::Attributes::SupportsConcurrentConnection::Id));
+        VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::GeneralCommissioning::Id,
+                                                Clusters::GeneralCommissioning::Attributes::Breadcrumb::Id));
+        VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::GeneralCommissioning::Id,
+                                                Clusters::GeneralCommissioning::Attributes::BasicCommissioningInfo::Id));
+        VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::GeneralCommissioning::Id,
+                                                Clusters::GeneralCommissioning::Attributes::RegulatoryConfig::Id));
+        VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::GeneralCommissioning::Id,
+                                                Clusters::GeneralCommissioning::Attributes::LocationCapability::Id));
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:2275]: A wildcard reead of general commissioning is probably small enough to get all you need and reduce number of paths there.

File: src/controller/CHIPDeviceController.cpp [hunk 5]
```diff
@@ -2189,46 +2194,166 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
...
+            VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::OperationalCredentials::Id,
+                                                    Clusters::OperationalCredentials::Attributes::Fabrics::Id));
+        }
+
+        // ICD Management
+        if (params.GetICDRegistrationStrategy() != ICDRegistrationStrategy::kIgnore)
+        {
+            VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::IcdManagement::Id,
+                                                    Clusters::IcdManagement::Attributes::FeatureMap::Id));
+        }
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:2305]: Here a wildcard would also help.

File: src/controller/CHIPDeviceController.cpp [hunk 6]
```diff
@@ -2189,46 +2194,166 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
...
+    {
+        mReadCommissioningInfoProgress = kReadProgressNoFurtherAttributes;
+    }
+
+    const auto timeout = MakeOptional(app::kExpectedIMProcessingTime); // TODO: Save timeout from PerformCommissioningStep?
+    SendCommissioningReadRequest(mDeviceBeingCommissioned, timeout, builder.paths(), builder.size());
+}
+
+namespace {
+void AccumulateErrors(CHIP_ERROR & acc, CHIP_ERROR err)
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:2336]: Suggest `AccumulateEarlierError`

File: src/controller/CHIPDeviceController.cpp [hunk 7]
```diff
@@ -2238,247 +2363,228 @@ void DeviceCommissioner::ParseCommissioningInfo()
...
-                info.network.thread.minConnectionTime = time;
+                ChipLogProgress(Controller, "NetworkCommissioning Features: has Ethernet. endpointid = %u", path.mEndpointId);
+                info.network.eth.endpoint = path.mEndpointId;
             }
-            else if (path.mEndpointId == info.network.eth.endpoint)
+            else
             {
-                info.network.eth.minConnectionTime = time;
+                ChipLogProgress(Controller, "NetworkCommissioning Features: no features.");
+                // TODO: Gross workaround for the empty feature map on all clusters. Remove.
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:2475]: That is a huge bug in all clusters. The right thing to do is to fix it.

File: src/controller/python/test/test_scripts/commissioning_failure_test.py [hunk 8]
```diff
@@ -106,7 +107,8 @@ async def main():
                       "Commissioning failure tests failed for simulated report failure on stage {}".format(testFailureStage))
 
     else:
-        for testFailureStage in range(3, 21):
+        # [kArmFailsafe, kConfigureTrustedTimeSource] (inclusive). TODO: https://github.com/project-chip/connectedhomeip/issues/36629
+        for testFailureStage in range(3, 20):
```
Comment (tcarmelveilleux) [inline src/controller/python/test/test_scripts/commissioning_failure_test.py:111]: This range change is odd. I also do not understand  the command

File: src/controller/CHIPDeviceController.cpp (hunk 4 above)
Comment (ksperling-apple) [reply src/controller/CHIPDeviceController.cpp:2275]: Yes, that's one of the optimizations I'm looking up for a follow-up. For ICD we can also do better I think, either by reading just the feature flag first and the rest during the second read (if any), or as a wildcard. But I wanted to keep those behavior changes separate from the refactoring.

File: src/controller/CHIPDeviceController.cpp (hunk 7 above)
Comment (ksperling-apple) [reply src/controller/CHIPDeviceController.cpp:2475]: Yeah, I'll have a look if this is still needed (the workaround is existing code).

File: src/controller/python/test/test_scripts/commissioning_failure_test.py (hunk 8 above)
Comment (ksperling-apple) [reply src/controller/python/test/test_scripts/commissioning_failure_test.py:111]: The range being iterated is kArmFailsafe to kConfigureTrustedTimeSource (inclusive), before and after. The numbers changed because I removed kReadCommissioningInfo2. I added the todo because we really should be exposing these as named constants to Python somehow.

File: src/python_testing/TC_CGEN_2_4.py [hunk 9]
```diff
@@ -50,14 +50,15 @@
...
-kSendNOC = 19
+# TODO: https://github.com/project-chip/connectedhomeip/issues/36629
+kArmFailsafe = 3
+kConfigRegulatory = 4
+kSendPAICertificateRequest = 9
+kSendDACCertificateRequest = 10
+kSendAttestationRequest = 11
+kSendOpCertSigningRequest = 14
+kSendTrustedRootCert = 17
+kSendNOC = 18
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/python_testing/TC_CGEN_2_4.py:61]: Indeed, that's why I filed that issue.

File: src/controller/CHIPDeviceController.cpp (hunk 3 above)
Comment (ksperling-apple) [reply src/controller/CHIPDeviceController.cpp]: Done

File: src/controller/CHIPDeviceController.cpp (hunk 6 above)
Comment (ksperling-apple) [reply src/controller/CHIPDeviceController.cpp:2345]: Hm, I'm not sure the "earlier" really adds clarity, I think I'll leave it as is for now.

File: src/controller/CHIPDeviceController.cpp (hunk 7 above)
Comment (cecille) [reply src/controller/CHIPDeviceController.cpp:2484]: I'm not sure this even happens any more - do you know if you're actually getting a blank feature map back?

**Review by cecille — COMMENTED** [review]
This generally seems better to me. Most of my comments are minor. +1 to Boris' question about cancellation though.

File: src/controller/CHIPDeviceController.cpp [hunk 10]
```diff
@@ -2189,46 +2194,175 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
...
+    bool exceeded() { return mCount > kCapacity; }
+    app::AttributePathParams * paths() { return mPaths; }
+
+    // Adds an attribute path if within the current window.
+    // Returns false if the available space has been exceeded.
+    template <typename... Ts>
+    bool AddAttributePath(Ts &&... args)
     {
-        err = ParseCommissioningInfo2(info);
+        if (mSkip > 0)
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/controller/CHIPDeviceController.cpp:2235]: Yes that’s correct, will add a note

File: src/controller/CHIPDeviceController.cpp [hunk 11]
```diff
@@ -2189,46 +2194,175 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
...
+                                                Clusters::IcdManagement::Attributes::ActiveModeThreshold::Id));
+    }();
+
+    VerifyOrDie(builder.size() > 0); // our logic is broken if there is nothing to read
+    if (builder.exceeded())
+    {
+        // Keep track of the number of attributes we have read already so we can resume from there.
+        auto progress = mReadCommissioningInfoProgress + builder.size();
+        VerifyOrDie(progress < kReadProgressNoFurtherAttributes);
+        mReadCommissioningInfoProgress = static_cast<decltype(mReadCommissioningInfoProgress)>(progress);
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/controller/CHIPDeviceController.cpp:2333]: Doing the math with uint8 is annoying due to promotion to int, so having the logic use size_t is convenient. At the same time I wanted to be a bit more frugal with what gets stored in the DeviceCommissioner permanently.

File: src/controller/CHIPDeviceController.cpp (hunk 1 above)
Comment (ksperling-apple) [reply src/controller/CHIPDeviceController.cpp]: Just to elaborate a bit more, cancellation calls `CancelCommissioningInteractions()`, we free the ReadClient (and now also the ClusterStateCache, this was missing before), and `OnDone` won't be called. So we won't get back into `ContinueReadingCommissioningInfo` or touch `mDeviceBeingCommissioned` in that case.

File: src/controller/CHIPDeviceController.cpp (hunk 7 above)
Comment (ksperling-apple) [reply src/controller/CHIPDeviceController.cpp:2490]: Yeah the current version of network-commissioning-server never returns a 0 feature map, I'll remove this.

File: src/darwin/Framework/CHIPTests/TestHelpers/MTRTestCase+ServerAppRunner.m [hunk 12]
```diff
@@ -61,7 +61,7 @@ + (NSTask *)doStartAppWithName:(NSString *)name arguments:(NSArray<NSString *> *
         @"--passcode",
         [NSString stringWithFormat:@"%llu", passcode.unsignedLongLongValue],
         @"--KVS",
-        [NSString stringWithFormat:@"/tmp/chip-%@-kvs%u", name, uniqueIndex],
+        [NSString stringWithFormat:@"/tmp/xctest-%d-chip-%@-kvs%u", getpid(), name, uniqueIndex],
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/TestHelpers/MTRTestCase+ServerAppRunner.m:64]: We should consider making a similar change for the outFile/errorFile bits below.  Followup is fine for that.

File: src/controller/CHIPDeviceController.h [hunk 13]
```diff
@@ -839,7 +839,9 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
     CommissioneeDeviceProxy * mDeviceInPASEEstablishment = nullptr;
 
     CommissioningStage mCommissioningStage = CommissioningStage::kSecurePairing;
-    bool mRunCommissioningAfterConnection  = false;
+    uint8_t mReadCommissioningInfoProgress = 0;
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.h:842]: Document that this tracks the number of attribute paths that had reads issued so far?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/darwin/Framework/CHIPTests/TestHelpers/MTRTestCase+ServerAppRunner.m (hunk 12 above)
Comment (ksperling-apple) [reply src/darwin/Framework/CHIPTests/TestHelpers/MTRTestCase+ServerAppRunner.m:64]: We could. For logs it's not really a problem to overwrite the existing file, but for the KVS we don't overwrite.

File: src/controller/CHIPDeviceController.cpp [hunk 14]
```diff
@@ -2259,6 +2259,7 @@ void DeviceCommissioner::ContinueReadingCommissioningInfo(const CommissioningPar
 {
     VerifyOrDie(mCommissioningStage == CommissioningStage::kReadCommissioningInfo);
 
+    // mReadCommissioningInfoProgress starts at 0 and counts the number of attributes we have read.
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:2262]: This is specifically the number of attribute paths we have issued reads for, not the number of attribute values we got back, due to wildcarding...

**Review by cecille — APPROVED** [review]
